tempesta_fw/http_parser.c

```diff
@@ -2251,6 +2312,116 @@ __req_parse_host(TfwHttpReq *req, unsigned char *data, size_t len)
 	return r;
 }
 
+/**
+ * Function for parsing 'Referer' header value. This is a partial
+ * copy from parsing procedure for request start line URI with some 
+ * simplifications according to RFC 7231 chapter 5.5.2.
+ */
+static int
+__req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len)
+{
+	int r = CSTR_NEQ;
+	unsigned char *off_ptr = data;
+	__FSM_DECLARE_VARS(hm);
+
+	__FSM_START(parser->_i_st) {
+
+	__FSM_STATE(Req_Uri) {
+		if (likely(c == '/')) {
+			__FSM_I_MOVE_offset_fn(Req_UriAbsPath, 1, off_ptr);
+		}
+		if (likely(__data_available(p, 7)
+			   && C4_INT_LCM(p, 'h', 't', 't', 'p')
+			   && *(p + 4) == ':'
+			   && *(p + 5) == '/'
+			   && *(p + 6) == '/'))
+		{
+			__FSM_I_MOVE_nofixup_n(Req_UriAuthorityStart, 7);
+		}
+		if (likely(TFW_LC(c) == 'h')) {
+			__FSM_I_MOVE_nofixup_n(Req_UriSchH, 1);
```

Comment:
Analogous to `__FSM_MOVE_nofixup()`, perhaps it makes sense to introduce `__FSM_I_MOVE_nofixup()` as well?